dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/action/validation/BaseProgramRuleActionValidator.java

```diff
@@ -193,4 +189,93 @@ private ProgramRuleActionValidationResult handleTrackedEntityAttribute(
 
     return ProgramRuleActionValidationResult.builder().valid(true).build();
   }
+
+  public List<ProgramStage> getRequiredProgramStages(
```

Comment:
We should follow this guide https://github.com/dhis2/wow-backend/blob/master/guides/code-guide.md#nullability--immutability and mark parameters as non null in the signature of the method, this way you will see that you can remove most of the null checks that you have in the code and it will be much more clean.